analyzers/src/SonarAnalyzer.CSharp/Rules/AvoidExcessiveInheritance.cs

```diff
@@ -36,89 +38,108 @@ namespace SonarAnalyzer.Rules.CSharp
+                    // See: https://github.com/dotnet/roslyn/issues/50989
+                    if (objectTypeInfo.Symbol == null || c.ContainingSymbol.Kind != SymbolKind.NamedType)
                     {
                         return;
                     }
 
-                    var thisTypeRootNamespace = GetRootNamespace(symbol);
+                    var thisTypeRootNamespace = GetRootNamespace(objectTypeInfo.Symbol);
 
-                    var baseTypesCount = symbol.GetSelfAndBaseTypes()
+                    var baseTypesCount = objectTypeInfo.Symbol.GetSelfAndBaseTypes()
                         .Skip(1) // remove the class itself
                         .TakeWhile(s => GetRootNamespace(s) == thisTypeRootNamespace)
                         .Select(nts => nts.OriginalDefinition.ToDisplayString())
-                        .TakeWhile(className => this.filteredClassesRegex.All(regex => !regex.IsMatch(className)))
+                        .TakeWhile(className => filteredClassesRegex.All(regex => !regex.IsMatch(className)))
                         .Count();
 
                     if (baseTypesCount > MaximumDepth)
                     {
-                        c.ReportDiagnosticWhenActive(Diagnostic.Create(rule, declaration.Identifier.GetLocation(),
-                            baseTypesCount, MaximumDepth));
+                        c.ReportDiagnosticWhenActive(Diagnostic.Create(rule, objectTypeInfo.Identifier.GetLocation(), objectTypeInfo.Name, baseTypesCount, MaximumDepth));
                     }
-
-                }, SyntaxKind.ClassDeclaration);
-        }
+                },
+                SyntaxKind.ClassDeclaration,
+                SyntaxKindEx.RecordDeclaration);
 
         private static string GetRootNamespace(ISymbol symbol)
         {
-            var namespaceString
-                = symbol.ContainingNamespace.ToDisplayString();
+            var namespaceString = symbol.ContainingNamespace.ToDisplayString();
 
-            var lastDotIndex = namespaceString.IndexOf(".");
+            var lastDotIndex = namespaceString.IndexOf(".", StringComparison.InvariantCulture);
             return lastDotIndex == -1
                 ? namespaceString
                 : namespaceString.Substring(0, lastDotIndex);
         }
 
-        private static Regex WilcardPatternToRegularExpression(string pattern)
+        private static Regex WildcardPatternToRegularExpression(string pattern)
         {
             var regexPattern = string.Concat("^", Regex.Escape(pattern).Replace("\\*", ".*"), "$");
             return new Regex(regexPattern, RegexOptions.Compiled);
         }
+
+        private class ObjectTypeInfo
+        {
+            public SyntaxToken Identifier { get; }
+
+            public INamedTypeSymbol Symbol { get; }
+
+            public string Name { get; }
+
+            public ObjectTypeInfo(SyntaxNode node, SemanticModel model)
```

Comment:
I am personally not a fan of creating such constructors that incorporate such kind of logic. If we will move this to some common space, may be lets create a factory?

Reply:
I want to keep things as simple as possible. I think we can create the factory when the need arises.